classes/handler/PKPHandler.php

```diff
@@ -331,7 +334,7 @@ public function authorize($request, &$args, $roleAssignments)
         if (!SessionManager::isDisabled()) {
             // Add user roles in authorized context.
             $user = $request->getUser();
-            if ($user instanceof \PKP\user\User || $request->getRouter() instanceof \PKP\core\APIRouter) {
+            if ($user instanceof \PKP\user\User || ($request->getRouter() instanceof \PKP\core\APIRouter && !$this->_isPublicApi)) {
```

Comment:
@bozana one problem with this approach is that it sets public/private at the Handler level, but often we want to add/remove authorization restrictions differently for each endpoint.

I'm going to ask @asmecher to take a look at this, because I think he's closer to the authorization stuff and may have a good idea for how to handle public endpoints. Alec, is there a good way to set up a public handler? How do the reader frontend pages do it?

Reply:
At a glance, we actually may not have any public read APIs yet (!) but there is an existing way around the check you've identified, @bozana. The reason the role based policy is currently required is this: at some point in the past we didn't require the policy, and some handlers didn't list any roles. This was easy to overlook and led to privileged features being exposed publicly. So role-based authorization is required as a safeguard (https://github.com/pkp/pkp-lib/issues/4024).

This includes a work-around function to explicitly disable this check when it's not intended ([`PKPHandler::markRoleAssignmentsChecked`](https://github.com/pkp/pkp-lib/blob/1bf2782e82ec9cdd582f786ba8c8b0f59eba528d/classes/handler/PKPHandler.php#L296)). You can call this from within the `authorize` function. See `lib/pkp/pages/submission/PKPSubmissionHandler.php` and `lib/pkp/pages/workflow/PKPWorkflowHandler.php` for examples.

Reply:
Oh, I have seen that function but I was not aware that it can be used for this... I will try... :+1:

Reply:
`markRoleAssignmentChecked` does not solve the problem -- the `UserRolesRequiredPolicy` here (at this line with the comment above) does require a user.
For SUSHI API however we do not need any user, the API should/can be public.

Thus, @asmecher, may I ask you again for a suggestion/opinion?

Reply:
This part of the code `($request->getRouter() instanceof \PKP\core\APIRouter)` comes from these changes:
https://github.com/pkp/pkp-lib/commit/efeec1fc7217be5c08a048273d0dc36bb80240f7#diff-634eadc948b55274d153a0d4b35fcd8529ffcc5041d33c9a68cc1fcc881b02ed

Reply:
I see two ways to solve this:
1. Remove that code `($request->getRouter() instanceof \PKP\core\APIRouter)` that requires a user for **all** APIs. Then we need to be **very careful** and double check that each API handler uses the correct policies (`UserRolesRequiredPolicy` and/or  `RoleBasedHandlerOperationPolicy`). 
Can maybe APIRouter here also mean something else?
2. Solve it as here in the PR: leave that code `($request->getRouter() instanceof \PKP\core\APIRouter)`, require the API handler to explicitly be defined as public (if any of that API's endpoints needs to be public), and if needed restrict other, not public endpoints of that API using a policy (e.g. `UserRolesRequiredPolicy` and/or  `RoleBasedHandlerOperationPolicy`).

@asmecher, what would you prefer i.e. do you see any other possibility?

Reply:
Hmm, I see what you mean.

I agree that solution 2 would be "safer" i.e. we would not have to review all existing handlers. However I would eventually like to move our API routing [from Slim to Laravel](https://github.com/pkp/pkp-lib/issues/7698), and the fewer "special cases" we have the better. We don't have so many API endpoints that manually adding `UserRolesRequiredPolicy` to each would be an impossible amount of work. And leaving the addition of `UserRolesRequiredPolicy` to each handler would mean they can selectively add it for only certain operations, which wouldn't be possible for a `$_isPublicApi` flag. So unfortunately I lean towards solution 1...

Reply:
@asmecher and/or @NateWr, if the API handlers use RoleBasedHandlerOperationPolicy do I need to add UserRolesRequiredPolicy too?

Reply:
I will add it, to be on the safe side... :)